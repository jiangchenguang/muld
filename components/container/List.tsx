import * as React from 'react';
import { View, ViewProps, AlignX, AlignY, horizontal, vertical } from './View';

// eslint-disable-next-line @typescript-eslint/no-empty-interface
interface ListProps extends ViewProps {
    alignX?: AlignX;
    alignY?: AlignY;
}

export const List: React.FC<React.PropsWithChildren<ListProps>> = (props) => {
    const { tag = 'ul', alignX, alignY, ...remainProps } = props;
    const style: React.CSSProperties = {
        display: 'flex',
        flexDirection: 'column',
        alignSelf: 'stretch',
        flex: '1 1 auto',
    };

    if (alignX) {
        style.alignItems = horizontal[alignX];
    }
    if (alignY) {
        style.justifyContent = vertical[alignY];
    }

    return (
        <View tag={tag} {...remainProps}>
            {props.children}
        </View>
    );
};
