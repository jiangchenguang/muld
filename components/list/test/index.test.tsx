import * as React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { render, cleanup, fireEvent } from '@testing-library/react';
import 'regenerator-runtime';
import List from '..';

const ListDemo = (): React.ReactElement => {
    const [loading, setLoading] = React.useState<boolean>(false);
    const [error, setError] = React.useState<boolean>(true);
    return (
        <List
            loading={loading}
            error={error}
            errorText="点击重新加载。"
            onLoad={(): void => {
                setLoading(true);
                setError(false);
            }}
        />
    );
};

describe('List', () => {
    afterEach(cleanup);
    it('loading', async () => {
        const { container } = render(<List loading={true} />);
        const loading = container.querySelector('.mul-list__loading');
        expect(loading).not.toBeNull();
    });

    it('error loaded, click error-text and reload', async () => {
        const { container } = render(<ListDemo />);
        const error = container.querySelector('.mul-list__error-text');
        fireEvent.click(error!);
        const loading = container.querySelector('.mul-list__loading');
        expect(loading).not.toBeNull();
    });

    test('finished', async () => {
        const { container } = render(
            <List loading={false} finished={true} finishedText="Finished" />,
        );
        const finished = container.querySelector('.mul-list__finished-text');
        expect(finished).not.toBeNull();
    });

    test('finished children', async () => {
        const { asFragment } = render(
            <List loading={false} finished={true} finishedText={<div>Custom Finished</div>} />,
        );
        expect(asFragment()).toMatchSnapshot();
    });

    test('error children', async () => {
        const { asFragment } = render(
            <List loading={false} error={true} errorText={<div>Custom Error</div>} />,
        );

        expect(asFragment()).toMatchSnapshot();
    });

    test('immediate check false', async () => {
        const { container } = render(<List immediateCheck={false} />);

        const loading = container.querySelector('.mul-list__loading');
        expect(loading).toBeNull();
    });

    test('check the direction props', () => {
        const { asFragment, rerender } = render(<List direction="up" />);

        expect(asFragment()).toMatchSnapshot();

        rerender(<List direction="down" />);

        expect(asFragment()).toMatchSnapshot();
    });
});
