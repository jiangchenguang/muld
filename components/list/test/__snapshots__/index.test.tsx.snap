// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List check the direction props 1`] = `
<DocumentFragment>
  <div
    class="sc-AxirZ jfMPHR mul-list"
    role="feed"
  >
    <div
      class="mul-list__placeholder"
    />
  </div>
</DocumentFragment>
`;

exports[`List check the direction props 2`] = `
<DocumentFragment>
  <div
    class="sc-AxirZ jfMPHR mul-list"
    role="feed"
  >
    <div
      class="mul-list__placeholder"
    />
  </div>
</DocumentFragment>
`;

exports[`List error children 1`] = `
<DocumentFragment>
  <div
    aria-busy="false"
    class="sc-AxirZ jfMPHR mul-list"
    role="feed"
  >
    <div
      class="mul-list__error-text"
    >
      <div>
        Custom Error
      </div>
    </div>
    <div
      class="mul-list__placeholder"
    />
  </div>
</DocumentFragment>
`;

exports[`List finished children 1`] = `
<DocumentFragment>
  <div
    aria-busy="false"
    class="sc-AxirZ jfMPHR mul-list"
    role="feed"
  >
    <div
      class="mul-list__finished-text"
    >
      <div>
        Custom Finished
      </div>
    </div>
    <div
      class="mul-list__placeholder"
    />
  </div>
</DocumentFragment>
`;
